File: r/tests/testthat/test-published-totals.R
## Checks eavs_load() against the totals the EAC publishes in its own reports.
##
## The figures come from metadata/reference_totals.csv, quoted from the EAC's
## Comprehensive Report with page provenance, so they are the agency's numbers
## rather than anything derived here. If sentinel decoding leaked a -99 into a
## sum, or a concept were mapped to the wrong year's code, these would move.
##
## Needs both the shared metadata and a populated cache, so it skips during
## R CMD check (which runs from a built copy with no metadata/ alongside) and
## runs under testthat::test_local() in a checkout. The Python suite has the
## same checks in py/tests/test_published_totals.py.

find_metadata <- function() {
  dir <- normalizePath(testthat::test_path("."), mustWork = FALSE)
  for (i in 1:6) {
    candidate <- file.path(dir, "metadata", "reference_totals.csv")
    if (file.exists(candidate)) {
      return(dirname(candidate))
    }
    parent <- dirname(dir)
    if (identical(parent, dir)) break
    dir <- parent
  }
  NULL
}

cached_years <- function(years) {
  cache <- eavs_cache_dir(create = FALSE)
  if (!dir.exists(cache)) {
    return(character())
  }
  vapply(years, function(y) {
    row <- eavs_manifest[eavs_manifest$survey == "eavs" &
                           eavs_manifest$format == "csv" &
                           eavs_manifest$year == y, ]
    nrow(row) == 1 && file.exists(file.path(cache, row$file_name))
  }, logical(1))
}

# A rate over only the jurisdictions that reported both sides. This is not a
# refinement: over all rows, uocava_counted / uocava_returned is 112% for 2024,
# because all 67 Alabama counties report the numerator and none the denominator.
pair_rate <- function(panel, numerator, denominator) {
  both <- !is.na(panel[[numerator]]) & !is.na(panel[[denominator]])
  100 * sum(panel[[numerator]][both]) / sum(panel[[denominator]][both])
}

# A voting-mode share of total participation among jurisdictions reporting that
# mode, which is the convention the EAC's own published shares follow.
mode_share <- function(panel, mode) {
  reported <- !is.na(panel[[mode]])
  100 * sum(panel[[mode]][reported]) / sum(panel$partic_total[reported], na.rm = TRUE)
}

metric_value <- function(panel, metric) {
  switch(
    metric,
    partic_total = sum(panel$partic_total, na.rm = TRUE),
    reg_active = sum(panel$reg_active, na.rm = TRUE),
    reg_forms_received = sum(panel$reg_forms_received, na.rm = TRUE),
    reg_confirmations_sent = sum(panel$reg_confirmations_sent, na.rm = TRUE),
    reg_removed_total = sum(panel$reg_removed_total, na.rm = TRUE),
    poll_workers_total = sum(panel$poll_workers_total, na.rm = TRUE),
    uocava_transmitted = sum(panel$uocava_transmitted, na.rm = TRUE),
    fwab_returned = sum(panel$fwab_returned, na.rm = TRUE),
    fwab_counted = sum(panel$fwab_counted, na.rm = TRUE),
    uocava_returned_rate = pair_rate(panel, "uocava_returned", "uocava_transmitted"),
    uocava_counted_rate = pair_rate(panel, "uocava_counted", "uocava_returned"),
    uocava_rejected_rate = pair_rate(panel, "uocava_rejected", "uocava_returned"),
    reg_new_valid_share = pair_rate(panel, "reg_new_valid", "reg_forms_received"),
    partic_in_person_ed_share = mode_share(panel, "partic_in_person_ed"),
    NULL
  )
}

test_that("published EAC report totals are reproduced", {
  meta <- find_metadata()
  skip_if(is.null(meta), "metadata/ not alongside the package")

  reference <- utils::read.csv(file.path(meta, "reference_totals.csv"),
                              stringsAsFactors = FALSE)
  years <- sort(unique(reference$year))
  have <- cached_years(years)
  skip_if_not(all(have), paste("needs cached EAVS csv for", toString(years[!have])))

  # Every citation needs a way to check it, or it sits in the file doing nothing.
  unimplemented <- unique(reference$metric[
    vapply(reference$metric, function(m) is.null(metric_value(NULL, m)), logical(1))
  ])
  expect_equal(unimplemented, character(0))

  panels <- lapply(years, function(y) eavs_load(y, quiet = TRUE))
  names(panels) <- as.character(years)

  for (i in seq_len(nrow(reference))) {
    row <- reference[i, ]
    got <- metric_value(panels[[as.character(row$year)]], row$metric)
    label <- sprintf("%d %s %s %s (%s)", row$year, row$metric, row$relation,
                     format(row$value, scientific = FALSE), row$source)
    if (row$relation == "gt") {
      expect_gt(got, row$value, label = label)
    } else if (row$relation == "lt") {
      expect_lt(got, row$value, label = label)
    } else {
      stop("unknown relation: ", row$relation)
    }
  }
})

test_that("rate denominators must be restricted to common reporters", {
  meta <- find_metadata()
  skip_if(is.null(meta), "metadata/ not alongside the package")
  skip_if_not(all(cached_years(2024)), "needs the 2024 file cached")

  panel <- eavs_load(2024, quiet = TRUE)
  naive <- 100 * sum(panel$uocava_counted, na.rm = TRUE) /
    sum(panel$uocava_returned, na.rm = TRUE)
  correct <- pair_rate(panel, "uocava_counted", "uocava_returned")

  # If someone "simplifies" pair_rate to a ratio of column sums, this fails.
  expect_gt(naive, 100)
  expect_gt(correct, 96)
  expect_lt(correct, 97)

  alabama <- panel[panel$state_abbr == "AL", ]
  expect_true(all(!is.na(alabama$uocava_counted)))
  expect_true(all(is.na(alabama$uocava_returned)))
})

File: py/tests/test_published_totals.py
"""Check tidyeavs against the totals the EAC publishes in its own reports.

This is the strongest external validation available: the figures in
``metadata/reference_totals.csv`` are quoted from the EAC's Comprehensive Report
with page provenance, so they are the agency's own numbers rather than anything
derived here. If the sentinel decoding leaked a -99 into a sum, or a concept were
mapped to the wrong year's code, these would move.

The report states figures as bounds ("over 158 million", "more than 96%"), so the
assertions are bounds too, which is what the source actually supports.

**Rate denominators are restricted to jurisdictions reporting both sides.** This
is not a refinement, it is the difference between right and wrong. Summing
``uocava_counted`` over every row and dividing by ``uocava_returned`` over every
row gives 112% for 2024, because all 67 Alabama counties report the numerator and
none report the denominator. On the common subset it is 96.4%, matching the
report. The same convention reproduces the report's Election Day share exactly.

Needs a populated cache, so skipped otherwise; see test_integration.py.
"""

from __future__ import annotations

import pandas as pd
import pytest

import tidyeavs
from tidyeavs.cache import cache_dir
from tidyeavs.download import manifest_lookup


def _reference() -> pd.DataFrame:
    return pd.read_csv(tidyeavs.metadata_dir() / "reference_totals.csv")


def _years_needed() -> list[int]:
    return sorted(_reference()["year"].unique())


def _cached(year: int) -> bool:
    root = cache_dir(create=False)
    if not root.is_dir():
        return False
    return (root / manifest_lookup(year).iloc[0]["file_name"]).is_file()


@pytest.fixture(scope="module")
def panels():
    years = _years_needed()
    missing = [y for y in years if not _cached(y)]
    if missing:
        pytest.skip(f"needs the {missing} file(s) cached")
    return {y: tidyeavs.load([y], quiet=True) for y in years}


def _pair_rate(panel: pd.DataFrame, numerator: str, denominator: str) -> float:
    """A rate over only the jurisdictions that reported both sides."""
    both = panel[numerator].notna() & panel[denominator].notna()
    return 100 * panel.loc[both, numerator].sum() / panel.loc[both, denominator].sum()


def _mode_share(panel: pd.DataFrame, mode: str) -> float:
    """A voting-mode share of total participation, among jurisdictions reporting
    that mode. This is the convention the EAC's own figures follow."""
    reported = panel[mode].notna()
    return 100 * panel.loc[reported, mode].sum() / panel.loc[reported, "partic_total"].sum()


# metric name -> how to compute it from one year's panel.
METRICS = {
    "partic_total": lambda p: p["partic_total"].sum(),
    "reg_active": lambda p: p["reg_active"].sum(),
    "reg_forms_received": lambda p: p["reg_forms_received"].sum(),
    "reg_confirmations_sent": lambda p: p["reg_confirmations_sent"].sum(),
    "reg_removed_total": lambda p: p["reg_removed_total"].sum(),
    "poll_workers_total": lambda p: p["poll_workers_total"].sum(),
    "uocava_transmitted": lambda p: p["uocava_transmitted"].sum(),
    "fwab_returned": lambda p: p["fwab_returned"].sum(),
    "fwab_counted": lambda p: p["fwab_counted"].sum(),
    "uocava_returned_rate": lambda p: _pair_rate(p, "uocava_returned", "uocava_transmitted"),
    "uocava_counted_rate": lambda p: _pair_rate(p, "uocava_counted", "uocava_returned"),
    "uocava_rejected_rate": lambda p: _pair_rate(p, "uocava_rejected", "uocava_returned"),
    "partic_in_person_ed_share": lambda p: _mode_share(p, "partic_in_person_ed"),
    "reg_new_valid_share": lambda p: _pair_rate(p, "reg_new_valid", "reg_forms_received"),
}


def test_every_reference_figure_has_an_implementation():
    """A citation with no way to check it would sit in the file doing nothing."""
    unimplemented = sorted(set(_reference()["metric"]) - set(METRICS))
    assert not unimplemented, f"reference_totals.csv metrics with no implementation: {unimplemented}"


@pytest.mark.parametrize(
    "row",
    [row for _, row in _reference().iterrows()],
    ids=lambda row: f"{row['year']}-{row['metric']}-{row['relation']}{row['value']:g}",
)
def test_matches_published_figure(panels, row):
    panel = panels[row["year"]]
    got = METRICS[row["metric"]](panel)
    expected = row["value"]

    if row["relation"] == "gt":
        assert got > expected, (
            f"{row['metric']} = {got:,.2f}, expected > {expected:,.2f}. "
            f"Source: {row['source']} — {row['note']}"
        )
    elif row["relation"] == "lt":
        assert got < expected, (
            f"{row['metric']} = {got:,.2f}, expected < {expected:,.2f}. "
            f"Source: {row['source']} — {row['note']}"
        )
    else:
        raise AssertionError(f"unknown relation {row['relation']!r}")


def test_naive_rate_denominators_are_wrong_and_this_is_why():
    """Guards the denominator convention itself, not just the result.

    Alabama 2024 reports uocava_counted for all 67 counties and uocava_returned
    for none, so a rate taken over all rows exceeds 100%. If someone "simplifies"
    _pair_rate to a plain ratio of column sums, this fails loudly.
    """
    if not _cached(2024):
        pytest.skip("needs the 2024 file cached")
    p = tidyeavs.load([2024], quiet=True)

    naive = 100 * p["uocava_counted"].sum() / p["uocava_returned"].sum()
    correct = _pair_rate(p, "uocava_counted", "uocava_returned")
    assert naive > 100, "expected the naive denominator to be visibly wrong"
    assert 96 < correct < 97

    alabama = p[p["state_abbr"] == "AL"]
    assert alabama["uocava_counted"].notna().all()
    assert alabama["uocava_returned"].isna().all()
